### PR TITLE
[CR] Fixed "Snow glare is active during night time #33413"

### LINKED ISSUE #33413: Snow glare is active during night time
You get perception penalties for snow glare during the night.

# Steps To Reproduce

Make a winter start character. Mine uses bifocal glasses, that could be a part of it. In the screenshot attached there's moonlight, but even in cases where there isnt any you still get the same penalties.

# Expected behavior

You shouldnt get snow glare during the night. The description speaks abount sunlight specifically.

# Screenshots

image
# Versions and configuration
 - OS: Windows 7
 - Game Version: 0.D-6740-g5b7c6f1 (tiles
 - Graphics version: Tiles
 - Mods loaded: No fungal monsters, Disabled NPC needs, Beta National Guard Camp

### PR DESCRIPTION
**Pull request by ipcyborg**
#### Summary
```SUMMARY: Bugfixes "Snow glare is active during night time #33413"```

#### Purpose of change
Fixes #33413

#### Describe the solution
I can't find any way to detect if there is snow on the tile.
So, assumed that winter means snow everywhere. At least it looks like it on the tileset. But the tile description still says: dirt, grass, etc.
Function `game::is_in_sunlight` considers only two type of weather as `sunlight`: `WEATHER_CLEAR` and `WEATHER_SUNNY`. Looks correct to me.

And `snow` weather means that it is snowing right now (not sunny or clear). So it should not lead to sun glare, right?

Refactored `weather_effect::glare` to avoid mistakes when adding other glare-related effects, like checking for specific mutations.
Now `glare` is called for both clear and sunny weather, but now for snow. To keep original behavior mostly intact, clear weather only add glare effect during winter (assuming snow everywhere).

Removed `weather_effect::snow_glare()`, can't find any usage of it.

Tested with various combinations:
- day / night
- inside / outside
- clear / sunny / show weather
- spring / winter
- with / without `CEPH_VISION` mutation

#### Describe alternatives you've considered
The same, but keeping the original behavior for `snow` weather (causing glare). Still seems wrong to me.

#### Additional context
Using only logic here, no actual IRL reference. 
Probably incorrect solution to remove glare from snow weather?

### CODE REVIEW

**Comment by ipcyborg** [comment]
Jenkins Rebuild

File: src/weather.cpp [hunk 1]
```diff
@@ -61,44 +61,37 @@ static bool is_player_outside()
...
-            if( g->u.has_trait( trait_CEPH_VISION ) ) {
-                g->u.add_env_effect( effect_glare, bp_eyes, 2, 2_turns );
-            } else {
-                g->u.add_env_effect( effect_glare, bp_eyes, 2, 1_turns );
-            }
-        }
+    if( season == WINTER ) {
+        //Winter snow glare: for both clear & sunny weather
+        effect = &effect_snow_glare;
+        dur = ( !g->u.has_effect( *effect ) ) ? 2_turns : 1_turns;
```
Comment (BevapDin) [inline src/weather.cpp]:
```suggestion
        dur = g->u.has_effect( *effect ) ? 1_turns : 2_turns;
```
Same below. It's just easier to read and avoids double negation.

File: src/weather.cpp [hunk 2]
```diff
@@ -61,44 +61,37 @@ static bool is_player_outside()
 /**
  * Glare.
  * Causes glare effect to player's eyes if they are not wearing applicable eye protection.
+ * @param intensity Level of sun brighthess (1 = clear, 2 = sunny)
```
Comment (BevapDin) [inline src/weather.cpp]: Seems like this should be an `enum`, as hinted by the explanation in the comment: there are only two possible values and you have given them names in the comment. Why not give those two values names in the code itself?

Comment (ipcyborg) [reply src/weather.cpp]: Fixed, thanks

**Comment by davidpwbrown** [comment]
Snowglare was removed in #28371 
Or, at least I thought it was.

**Comment by ipcyborg** [comment]
> Snowglare was removed in #28371

That commit was about removing `glare` function from `weather_effect::snow`. 

>Snowglare logic was inconistent.

Totally agree with it. So I just did the same, removed `glare` from `weather_effect::snow`. But also added glare for sunny & clear weather in winter. 
It is not perfect, but I think it is more logical that the current state.

Alternative: just remove snowglare completely.

**Comment by davidpwbrown** [comment]
> > Snowglare was removed in #28371
> 
> That commit was about removing `glare` function from `weather_effect::snow`.
> 
> > Snowglare logic was inconistent.
> 
> Totally agree with it. So I just did the same, removed `glare` from `weather_effect::snow`. But also added glare for sunny & clear weather in winter.
> It is not perfect, but I think it is more logical that the current state.
> 
> Alternative: just remove snowglare completely.

Thats weird then, unless im missing something. 

`void weather_effect::snow()
{
    wet_player( 10 );
    glare( true );
}`  glare(true) should have been removed ages ago, and yet it is somehow back there!

**Comment by ipcyborg** [comment]
>the glare( true ) should have been removed ages ago, and yet it is somehow back there!

Looks like that PR #28371 was never committed, is was just closed:

>Merged to master, then merged and reverted to development.
>kevingranade closed this on Feb 25

However, the same change is also included in this PR.

**Pull request merged**